lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManager.java

```diff
@@ -212,6 +230,45 @@ private synchronized boolean haveAllExpectedEvaluatorsReported() {
     return true;
   }
 
+  /**
+   * Sets the driver restart status to be completed if not yet set and notifies the restart completed event handlers.
+   */
+  private synchronized void onDriverRestartCompleted() {
+    if (this.state != DriverRestartState.COMPLETED) {
+      final Set<String> outstandingEvaluatorIds = getOutstandingEvaluatorsAndMarkExpired();
+      driverRuntimeRestartManager.informAboutEvaluatorFailures(outstandingEvaluatorIds);
+
+      this.state = DriverRestartState.COMPLETED;
+      final DriverRestartCompleted driverRestartCompleted = new DriverRestartCompleted(System.currentTimeMillis());
+
+      for (final EventHandler<DriverRestartCompleted> serviceRestartCompletedHandler
```

Comment:
Doing this in a loop is probably OK. If it ever becomes a performance issue, we can parallelize it.